File: app/src/main/java/quickbeer/android/data/stores/UserStore.java
/*
 * The MIT License
 *
 * Copyright (c) 2013-2016 reark project contributors
 *
 * https://github.com/reark/reark/graphs/contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package quickbeer.android.data.stores;

import android.content.ContentResolver;
import android.support.annotation.NonNull;

import com.google.gson.Gson;

import polanski.option.Option;
import quickbeer.android.Constants;
import quickbeer.android.data.pojos.User;
import quickbeer.android.data.stores.cores.UserStoreCore;

/**
 * Store to keep user data.
 */
public class UserStore extends StoreBase<Integer, User, Option<User>> {

    public UserStore(@NonNull final ContentResolver contentResolver,
                     @NonNull final Gson gson) {
        super(new UserStoreCore(contentResolver, gson),
              item -> Constants.DEFAULT_USER_ID,
              Option::ofObj,
              Option::none);
    }
}

File: app/src/main/java/quickbeer/android/data/stores/UserStore.kt
/**
 * This file is part of QuickBeer.
 * Copyright (C) 2017 Antti Poikela <antti.poikela@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package quickbeer.android.data.stores

import android.content.ContentResolver
import com.google.gson.Gson
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Function
import io.reark.reark.data.stores.DefaultStore.GetEmptyValue
import io.reark.reark.data.stores.DefaultStore.GetIdForItem
import io.reark.reark.data.stores.DefaultStore.GetNullSafe
import polanski.option.Option
import quickbeer.android.Constants
import quickbeer.android.data.pojos.User
import quickbeer.android.data.stores.cores.CachingStoreCore
import quickbeer.android.data.stores.cores.UserStoreCore

/**
 * Store to keep user data.
 */
class UserStore(contentResolver: ContentResolver, gson: Gson) :
    StoreBase<Int, User, Option<User>>(
        CachingStoreCore(
            UserStoreCore(contentResolver, gson),
            Function { Constants.DEFAULT_USER_ID },
            BiFunction { v1, v2 -> User.merge(v1, v2) }),
        GetIdForItem { Constants.DEFAULT_USER_ID },
        GetNullSafe { Option.ofObj(it) },
        GetEmptyValue { Option.none<User>() })
